rust/xaynet-analytics/src/controller.rs

```diff
@@ -0,0 +1,158 @@
+        let combiner = DataCombiner {};
+        let sender = Sender {};
+        Ok(AnalyticsController {
+            db,
+            is_charging,
+            is_connected_to_wifi,
+            last_time_data_sent,
+            analytics_event_repo,
+            controller_data_repo,
+            screen_route_repo,
+            combiner,
+            sender,
+            send_data_frequency: Duration::hours(Self::SEND_DATA_FREQUENCY_HOURS),
+        })
+    }
+
+    pub fn dispose(self) -> Result<(), Error> {
+        self.db.dispose()
+    }
+
+    pub fn save_analytics_event(&self, event_bytes: &[u8]) -> Result<(), Error> {
+        let event = AnalyticsEvent::read(event_bytes);
+        self.check_for_new_screen_route(&event)?;
+        self.analytics_event_repo.add(&event, &self.db)
+    }
+
+    pub fn change_connectivity_status(&mut self) {
+        self.is_connected_to_wifi = !self.is_connected_to_wifi
+    }
+
+    pub fn change_state_of_charge(&mut self) {
+        self.is_charging = !self.is_charging
+    }
+
+    pub fn maybe_send_data(&mut self) -> Result<(), Error> {
+        if self.should_send_data() {
+            self.send_data()
+        } else {
+            Ok(())
+        }
+    }
+
+    fn check_for_new_screen_route(&self, event: &AnalyticsEvent) -> Result<(), Error> {
+        match event.screen_route {
+            Some(screen_route) => self.add_screen_route_if_new(screen_route),
+            None => Ok(()),
+        }
+    }
+
+    fn add_screen_route_if_new(&self, screen_route: &ScreenRoute) -> Result<(), Error> {
+        if !self
+            .screen_route_repo
+            .get_all(&self.db)?
+            .contains(screen_route)
+        {
+            self.screen_route_repo.add(screen_route, &self.db)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn get_last_time_data_sent(
+        repo: &ControllerDataRepo,
+        db: &IsarDb,
+    ) -> Result<Option<DateTime<Utc>>, Error> {
+        match repo.get_all(db)?.last() {
+            Some(data) => Ok(Some(data.time_data_sent)),
+            None => Ok(None),
+        }
+    }
+
+    fn should_send_data(&self) -> bool {
+        self.can_send_data() && !self.did_send_already_in_this_period()
+    }
+
+    fn can_send_data(&self) -> bool {
+        self.is_charging && self.is_connected_to_wifi
+    }
+
+    fn did_send_already_in_this_period(&self) -> bool {
+        if self.last_time_data_sent.is_none() {
+            false
+        } else {
+            let tomorrow = Utc::now() + Duration::days(1);
+            let midnight_after_current_time: DateTime<Utc> = DateTime::from_utc(
+                NaiveDate::from_ymd(tomorrow.year(), tomorrow.month(), tomorrow.day())
+                    .and_hms(0, 0, 0),
+                Utc,
+            );
+            let start_of_curret_period = midnight_after_current_time - self.send_data_frequency;
+            self.last_time_data_sent.unwrap() < start_of_curret_period
```

Comment:
uhm, maybe, not sure i completely understand. if it's OK with the upcoming unit tests i'll check out different scenarios

Reply:
to elaborate a bit more on the example: suppose `send_data_frequency` is set to 12 hours and `Utc::now()` gives `today:06:00`. then `midnight_after_current_time` is `tomorrow:00:00` and therefore `start_of_curret_period` is `today:12:00`. but instead `start_of_curret_period` should be `today:00:00`, because we are still in the time frame `[today:00:00, today:12:00]`.

now we can send data arbitrary many times in the current period and `did_send_already_in_this_period()` will always return `false`.

Reply:
thanks for spotting the typo! i hear you both, with the unit tests i'll check both method name and logic. it's very likely there's some bits missing